src/pybind/mgr/k8sevents/module.py

```diff
@@ -0,0 +1,1072 @@
+# Integrate with the kubernetes events API. 
+# This module sends events to Kubernetes, and also captures/tracks all events
+# in the rook-ceph namespace so kubernetes activity like pod restarts,
+# imagepulls etc can be seen from within the ceph cluster itself.
+#
+# To interact with the events API, the mgr service to access needs to be 
+# granted additional permissions
+# e.g. kubectl -n rook-ceph edit clusterrole rook-ceph-mgr-cluster-rules
+#
+# These are the changes needed;
+# - apiGroups:
+#   - ""
+#   resources:
+#   - events
+#   verbs:
+#   - create
+#   - patch
+#   - list
+#   - get
+#   - watch
+
+
+import os
+import sys
+import time
+import json
+import logging
+import threading
+
+from datetime import datetime, timedelta, timezone
+from urllib3.exceptions import MaxRetryError
+from collections import OrderedDict
+
+import rados
+from mgr_module import MgrModule
+from ceph_argparse import run_in_thread
+
+try:
+    import queue
+except ImportError:
+    # python 2.7.5
+    import Queue as queue
+finally:
+    # python 2.7.15 or python3
+    event_queue = queue.Queue()
+
+try:
+    from kubernetes import client, config, watch
+    from kubernetes.client.rest import ApiException
+except ImportError:
+    kubernetes_imported = False
+    client = None
+    config = None
+else:
+    kubernetes_imported = True
+    # Apply the kubernetes config from the cluster 
+    config.load_incluster_config()
+
+    # The watch.Watch.stream method can provide event objects that have involved_object = None
+    # which causes an exception in the generator. A workaround is discussed for a similar issue
+    # in https://github.com/kubernetes-client/python/issues/376 which has been used here
+    # pylint: disable=no-member
+    from kubernetes.client.models.v1_event import V1Event
+    def local_involved_object(self, involved_object):
+        if involved_object is None:
+            involved_object = client.V1ObjectReference(api_version="1")
+        self._involved_object = involved_object
+    V1Event.involved_object = V1Event.involved_object.setter(local_involved_object)
+    #setattr(V1Event, 'involved_object', property(fget=V1Event.involved_object.fget, fset=local_involved_object))
+
+DEFAULT_LOG_LEVEL = 'info'
+DEFAULT_CONFIG_CHECK_SECS = 10
```

Comment:
10s is a short default interval if we're querying and checking the status of a number of objects. What about a default of 60s? That would also match rook's default interval for checking ceph health status.

Reply:
It's not that short for a manager module since the data is directly available in the mgr maps - i.e. we're not sending any commands to ceph. 10s or even 5s is what we use for prometheus scrapes that get data in the same way.